### PR TITLE
[cmake] [bugfix] provide correct args for target_include_directories call

### PR DESCRIPTION
**Pull request by linev**
One has to exclude "-I" prefix which was required for direct
command for dicitionary generation. 
When calling `target_include_directories()`, it has to be removed. 
Otherwise list of includes for dictionary compilation looks:
```
CXX_INCLUDES = -I/home/linev/git/root6/io/io/-I/home/linev/git/root6 -I/home/linev/git/root6/io/io/-I/home/linev/build/root6/etc/cling -I/home/linev/git/root6/io/io/-I/home/linev/build/root6/include -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/io/io -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/io/io/res -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/clib/res -I/home/linev/git/root6/io/io/-I/usr/include -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/base/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/base/v7/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/clib/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/cont/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/foundation/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/macosx/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/unix/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/winnt/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/clingutils/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/meta/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/gui/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/textinput/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/thread/inc -I/home/linev/git/root6/io/io/-I/home/linev/git/root6/core/thread -I/home/linev/git/root6/io/io/res -I/home/linev/build/root6/include -I/home/linev/git/root6/core/clib/res
```

### CODE REVIEW

**Comment by linev** [comment]
`${includedirs}` has more entries. 
It extract includes from `DEPENDENCIES` plus initial values in lines 372..379

**Comment by linev** [comment]
@amadio #5170 is full PR, where I completely exclude source dir from include paths.
This PR could be recognized as bugfix and also can be applied for 6.20 and 6.18 patch branches

**Comment by vgvassilev** [comment]
Could you expand a little more the commit message and the PR description. I fail to capture the intent.

In general, ROOT_GENERATE_DICTIONARY is very complex and needs to be simplified and refactored, unless this is fixing a real issue I'd be reluctant to make it more complex than it already is.

**Comment by amadio** [comment]
We *do* want to use the source directory includes when possible. So I don't agree that we should exclude those includes. What we have is already almost as good as we can have. We just need to fix the issue with lack of space between includes, and generate headers in `${CMAKE_CURRENT_BINARY_DIR}` instead of `${CMAKE_BINARY_DIR}/include`. See PRs #4108, #4281, and #4299 for more information on the subject. Note that many of those missing dependencies from #4299 are still not fixed, because it was decided that each person would take care of their parts of the code...

**Comment by linev** [comment]
@vgvassilev 
This is a bugfix - wrong argument is provided to `target_include_directories()` cmake function.
Result of produce CXX flags shown in the beginning.

**Comment by linev** [comment]
In other PR @amadio made comment:
> We do want to use the source directory includes when possible

But there is problem with several generated headers like `RConfigure.h` - is it the only one?
Only because of it we have to add `$ROOTSYS/include` in ROOT build paths.
Ending in situation, that every header available twice - via source dir and via '$ROOTSYS/include` dir. This should be avoided in any case. 

I propose a solution - generate `RConfigure.h` in special directory like `${CMAKE_BINARY_DIR}/generated_includes` and copy it into `${CMAKE_BINARY_DIR}/includes`. And in the list of include paths for ROOT libraries only source dirs and `${CMAKE_BINARY_DIR}/generated_includes` should be used. 
Do we agree on this? @amadio @vgvassilev @Axel-Naumann @oshadura

**Comment by Axel-Naumann** [comment]
> We do want to use the source directory includes when possible.

That's fine, but how's that connected to this PR (which removes the `-I` prefix from the list elements)? This PR is fixing a bug and doesn't preclude other, independent improvements.

File: cmake/modules/RootMacros.cmake [hunk 1]
```diff
@@ -553,8 +553,17 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
     target_compile_definitions(${dictionary} PRIVATE
       ${definitions} $<TARGET_PROPERTY:${ARG_MODULE},COMPILE_DEFINITIONS>)
 
+    # remove all -I prefixes from list of include dirs
+    # otherwise generated make files will look like:
+    # CXX_INCLUDES = -I/home/user/git/root6/io/io/-I/home/user/git/root6 ...
```
Comment (vgvassilev) [inline cmake/modules/RootMacros.cmake]: Sorry for being slow but I still fail to understand what is wrong with `-I/home/user/git/root6/io/io/-I/home/user/git/root6`. Is it the fact that there is no `inc` suffix or the fact that there is no space between `-I` or something else?

Comment (linev) [reply cmake/modules/RootMacros.cmake]: While cmake try to use directory `/home/user/git/root6/io/io/-I/home/user/git/root6` as includes and for dependency generation.

Comment (Axel-Naumann) [reply cmake/modules/RootMacros.cmake]: As in: `target_include_directories` expects a list of directories, but before Sergey's change got a list of `-I/some/dir`s. Not good.

Comment (amadio) [reply cmake/modules/RootMacros.cmake]: The question to be asked should be why is there no space in the list that was constructed in the first place? It's usually a side-effect from mixing `list($LIST APPEND ...)` and `set(LIST $LIST $EXTRA_STUFF)`. We should first try to fix this underlying issue in my opinion.

Comment (linev) [reply cmake/modules/RootMacros.cmake]: > The question to be asked should be why is there no space in the list

Problem that path starts with "-I" and cmake just prepend current source path. 
In no case one should call `target_include_directories()` providing such dir names. 
It is bug and I try fix it

**Comment by oshadura** [comment]
@linev I tried to reproduce your issue on my computer and I have next output:
`./core/CMakeFiles/G__Core.dir/flags.make:CXX_INCLUDES = -I/home/oksana/CERN_sources/root/core/base/inc -I/home/oksana/CERN_sources/root/core/clib/inc -I/home/oksana/CERN_sources/root/core/cont/inc -I/home/oksana/CERN_sources/root/core/foundation/inc -I/home/oksana/CERN_sources/root/core/macosx/inc -I/home/oksana/CERN_sources/root/core/unix/inc -I/home/oksana/CERN_sources/root/core/winnt/inc -I/home/oksana/CERN_sources/root/core/clingutils/inc -I/home/oksana/CERN_sources/root/core/meta/inc -I/home/oksana/CERN_sources/root/core/gui/inc -I/home/oksana/CERN_sources/root/core/textinput/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/builds/etc/cling -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/builds/include -I/home/oksana/CERN_sources/root/core/-I/usr/include -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/base/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/clib/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/cont/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/foundation/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/macosx/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/unix/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/winnt/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/clingutils/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/meta/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/gui/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/textinput/inc -I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core -I/home/oksana/CERN_sources/root/builds/include -I/home/oksana/CERN_sources/root -I/home/oksana/CERN_sources/root/builtins/xxhash `

**Comment by linev** [comment]
@oshadura 

You have exactly my problem! Corrupted entries are at the end of your list:
```
-I/home/oksana/CERN_sources/root/core/-I/home/oksana/CERN_sources/root/core/gui/inc
```

**Comment by linev** [comment]
I we agree to merge it - this PR should be applied only to release branches.
For the master #5181 should fix this the problem

**Comment by oshadura** [comment]
With your PR: `./core/CMakeFiles/G__Core.dir/flags.make:CXX_INCLUDES = -I/home/oksana/CERN_sources/root/core/base/inc -I/home/oksana/CERN_sources/root/core/clib/inc -I/home/oksana/CERN_sources/root/core/cont/inc -I/home/oksana/CERN_sources/root/core/foundation/inc -I/home/oksana/CERN_sources/root/core/macosx/inc -I/home/oksana/CERN_sources/root/core/unix/inc -I/home/oksana/CERN_sources/root/core/winnt/inc -I/home/oksana/CERN_sources/root/core/clingutils/inc -I/home/oksana/CERN_sources/root/core/meta/inc -I/home/oksana/CERN_sources/root/core/gui/inc -I/home/oksana/CERN_sources/root/core/textinput/inc -I/home/oksana/CERN_sources/root -I/home/oksana/CERN_sources/root/builds/etc/cling -I/home/oksana/CERN_sources/root/builds/include -I/home/oksana/CERN_sources/root/core -I/home/oksana/CERN_sources/root/builtins/xxhash 
`
@vgvassilev @amadio what do you think, can we try find an agreement here? Personally I think it is at least a quality improvement.

**Comment by vgvassilev** [comment]
It is your call, I like the fact that we do not duplicate include paths and we make things less. I do not like the fact it happens at the cost of introducing yet another variable. If there is no way to do it without it, then I'd be okay to merge this PR.

**Comment by linev** [comment]
My changes less invasive and just fixing a bug. In #5181 I trying to solve same problem without duplication of variables and for this purpose using `INCLUDE` argument

**Comment by linev** [comment]
We didnot come to any solution.
This code will be changed anyway after #5181, older ROOT branches will stay with a bug

**Pull request closed without merging**